syntax = "proto3";

package csce438;

import "google/protobuf/timestamp.proto";

// The messenger service definition.
service SNSService{

  rpc Login (Request) returns (Reply) {}
  rpc List (Request) returns (ListReply) {}
  rpc Follow (Request) returns (Reply) {}
  rpc UnFollow (Request) returns (Reply) {}
  // Bidirectional streaming RPC
  rpc Timeline (stream Message) returns (stream Message) {} 
}

//The Coordination Service definition
service SNSCoord{
  rpc ClusterSpawn (ClusterInfo) returns (Reply) {} //Server sends this to coord
  rpc GetConnection (JoinReq) returns (ClusterInfo) {} //Client sends this to coord
  rpc Gucci (HrtBt) returns (HrtBt) {} //Are you stil there? (´･_･`)
}


message ListReply {
  repeated string all_users = 1;
  repeated string followers = 2;
}

message Request {
  string username = 1;
  repeated string arguments = 2;
}

message Reply {
  string msg = 1;
}

message Message {
  //Username who sent the message
  string username = 1;
  //Message from the user
  string msg = 2;
  //Time the message was sent
  google.protobuf.Timestamp timestamp = 3;
}

message ClusterInfo {
  //The address of the machine running this clutster
  string addr = 1;
  //The port of the machine running this clutster
  string port = 2;
  //The the server ID of this cluster
  int32 id = 3;
}

//Signifys that a user of this ID wants to join a cluster
message JoinReq {
  int32 id = 1;
}

//A little heartbeat ◠‿◠, prolly subject to change
message HrtBt {
  int32 id = 1;
  bool master = 2;
}